### PR TITLE
Feature/tarantool api

### PR DESCRIPTION
**Pull request by roman-kozachenko**

### CODE REVIEW

File: src/iproto/Tuple.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
+﻿namespace iproto
+{
+    public interface ITuple { }
+    public class Tuple<T1> : ITuple
```
Comment (aensidhe) [inline src/iproto/Tuple.cs]: You need equality and hashcode members for these classes.

File: src/iproto/Tuple.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            Item1 = item1;
+        }
+        public T1 Item1 { get; }
+    }
+    public class Tuple<T1, T2> : ITuple
+    {
+        public Tuple(T1 item1, T2 item2)
+        {
+            Item1 = item1;
+            Item1 = item1; Item2 = item2;
```
Comment (aensidhe) [inline src/iproto/Tuple.cs]: Something went wrong here.

File: src/tarantool-client/Converters/TupleConverter.cs [hunk 3]
```diff
@@ -47,21 +49,21 @@ public void Write(Tuple<T1, T2> value, IMsgPackWriter writer, MsgPackContext con
             t2Converter.Write(value.Item2, writer, context);
         }
 
-        public Tuple<T1, T2> Read(IMsgPackReader reader, MsgPackContext context, Func<Tuple<T1, T2>> creator)
+        public iproto.Tuple<T1, T2> Read(IMsgPackReader reader, MsgPackContext context, Func<iproto.Tuple<T1, T2>> creator)
         {
             var t1Converter = context.GetConverter<T1>();
             var t2Converter = context.GetConverter<T2>();
 
             var item1 = t1Converter.Read(reader, context, null);
```
Comment (aensidhe) [inline src/tarantool-client/Converters/TupleConverter.cs]: You need to write array header here.

**Comment by aensidhe** [comment]
Please, commit generator here too so other people could contribute to it.

File: src/iproto/Tuple.cs (hunk 1 above)
Comment (roman-kozachenko) [inline src/iproto/Tuple.cs]: Added (to generation snippet also).

File: src/iproto/Tuple.cs (hunk 2 above)
Comment (roman-kozachenko) [inline src/iproto/Tuple.cs]: Yes, my fault, fixed.

File: src/tarantool-client/Converters/TupleConverter.cs (hunk 3 above)
Comment (roman-kozachenko) [inline src/tarantool-client/Converters/TupleConverter.cs]: Fixed. Forgot to merge changes from master.

File: .idea/modules.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectModuleManager">
```
Comment (aensidhe) [inline .idea/modules.xml]: Add .idea to .gitignore

File: _T4Awesome/config.t4ac [hunk 5]
```diff
@@ -0,0 +1 @@
+{"PCID":"b79cb3f2-b3aa-411c-a7c3-4813e5da7344","TemplateSearchPaths":[],"Version":"0.0.4","DBConnections":[],"DefaultTemplateLanguage":"C#"}
```
Comment (aensidhe) [inline _T4Awesome/config.t4ac]: What is this file?

File: src/tarantool-client/Converters/EnumConverter.cs [hunk 6]
```diff
@@ -18,106 +39,90 @@ static EnumConverter()
             }
         }
 
-        public void Write(T value, IMsgPackWriter writer, MsgPackContext context)
+        public void Write(T value, IMsgPackWriter writer)
```
Comment (aensidhe) [inline src/tarantool-client/Converters/EnumConverter.cs]: We can cache Func<T> to eliminate these lots of ifs

File: src/tarantool-client/Converters/ReflectionConverter.cs [hunk 7]
```diff
@@ -7,30 +7,37 @@
...
 
-using TarantoolDnx.MsgPack;
+using MsgPack.Light;
 
 namespace tarantool_client.Converters
 {
     public class ReflectionConverter : IMsgPackConverter<object>
     {
-        public void Write(object value, IMsgPackWriter writer, MsgPackContext context)
+        private MsgPackContext _context;
```
Comment (aensidhe) [inline src/tarantool-client/Converters/ReflectionConverter.cs]: Why do we need this in production code?

File: _T4Awesome/config.t4ac (hunk 5 above)
Comment (roman-kozachenko) [inline _T4Awesome/config.t4ac]: As you wrote:

> Please, commit generator here too so other people could contribute to it.

I use T4Awesome tool to generate Tuple classes. That file is its config.

File: src/tarantool-client/Converters/ReflectionConverter.cs (hunk 7 above)
Comment (roman-kozachenko) [inline src/tarantool-client/Converters/ReflectionConverter.cs]: Actually, you should ask that question. That converter can be useful for example in non-typed Select method overload. But if that scenario will not be used, we can remove that converter, and some methods overloadings.

File: .idea/modules.xml (hunk 4 above)
Comment (roman-kozachenko) [inline .idea/modules.xml]: Done.

File: src/tarantool-client/Converters/EnumConverter.cs (hunk 6 above)
Comment (roman-kozachenko) [inline src/tarantool-client/Converters/EnumConverter.cs]: Done.

**Pull request merged**